rust/kernel/chrdev.rs

```diff
@@ -1,100 +1,106 @@
 // SPDX-License-Identifier: GPL-2.0
 
 use core::convert::TryInto;
-use core::mem;
-use core::ops::Range;
-
-use alloc::boxed::Box;
-use alloc::vec;
-use alloc::vec::Vec;
+use core::marker::PhantomPinned;
+use core::mem::MaybeUninit;
+use core::pin::Pin;
 
 use crate::bindings;
 use crate::c_types;
 use crate::error::{Error, KernelResult};
 use crate::file_operations;
 use crate::types::CStr;
 
-pub fn builder(name: CStr<'static>, minors: Range<u16>) -> KernelResult<Builder> {
-    Ok(Builder {
-        name,
-        minors,
-        file_ops: vec![],
-    })
+struct RegistrationInner<const N: usize> {
+    dev: bindings::dev_t,
+    used: usize,
+    cdevs: [MaybeUninit<bindings::cdev>; N],
+    _pin: PhantomPinned,
 }
 
-pub struct Builder {
+/// chrdev registration. May contain up to a fixed number (`N`) of devices.
+/// Must be pinned.
+pub struct Registration<const N: usize> {
     name: CStr<'static>,
-    minors: Range<u16>,
-    file_ops: Vec<&'static bindings::file_operations>,
+    minors_start: u16,
+    this_module: &'static crate::ThisModule,
+    inner: Option<RegistrationInner<N>>,
 }
 
-impl Builder {
-    pub fn register_device<T: file_operations::FileOperations>(mut self) -> Builder {
-        if self.file_ops.len() >= self.minors.len() {
-            panic!("More devices registered than minor numbers allocated.")
+impl<const N: usize> Registration<{ N }> {
+    pub fn new(
+        name: CStr<'static>,
+        minors_start: u16,
+        this_module: &'static crate::ThisModule,
+    ) -> Self {
+        Registration {
+            name,
+            minors_start,
+            this_module,
+            inner: None,
         }
-        self.file_ops
-            .push(&file_operations::FileOperationsVtable::<T>::VTABLE);
-        self
     }
 
-    pub fn build(self, this_module: &'static crate::ThisModule) -> KernelResult<Registration> {
-        let mut dev: bindings::dev_t = 0;
-        let res = unsafe {
-            bindings::alloc_chrdev_region(
-                &mut dev,
-                self.minors.start.into(),
-                self.minors.len().try_into()?,
-                self.name.as_ptr() as *const c_types::c_char,
-            )
-        };
-        if res != 0 {
-            return Err(Error::from_kernel_errno(res));
+    pub fn register<T: file_operations::FileOperations>(self: Pin<&mut Self>) -> KernelResult<()> {
+        // SAFETY: we must ensure that we never move out of `this`.
+        let this = unsafe { self.get_unchecked_mut() };
+        if this.inner.is_none() {
```

Comment:
I guess this is done to simplify usage for callers, right? i.e. instead of having another method to allocate the region (or a typestate pattern) to do this once.

Reply:
That's right.

Reply:
Can we do it as part of `new()` and then no need to check on drop?

Reply:
Nope -- it's not pinned in new, and this memory can't move.

Reply:
But we could return the pinned one in `new()`, no? i.e. we know it is going to be heap pinning, and we can always impose the type of pinned pointer they need to use. Or better, we abstract that inside `Registration`.

Reply:
If we enforced heap pinning we could, but part of the objective here is not to require it to be on the heap, to let the caller provide the memory where it'd live.

Reply:
Yeah, but how does the module use it (without an allocation)? IIUC, we can create a pin manually every time we need to use the `Registration`, e.g.

    struct M {
        dev: chrdev::Registration<1>,
    }

    // some module function used somewhere
    fn f(m: &mut M) {
        // SAFETY: we will not move it
        let p = unsafe { Pin::new_unchecked(&mut m.dev) };

        // use it
        p.register::<...>();
    }

but how do we do it during module initialization? We need to move the memory to return `KernelResult<M>`, so we cannot pin it before that. Do we need to create a "deferred module initialization" call or should we change `init()` to pass it the uninitialized/default-initialized `M` memory from the static instead of asking to return a new object?

I feel I am missing something obvious, so apologies if that is the case.

Reply:
Yes, I'd expect callers to still need to use it `Box::pin()`. I do think eventually we'll want to do some sort of two-stage module initialization to allow for this, but I don't think it should block this.

You can see the same pattern in miscdev, its got the `new_pinned` API to handle common use cases, do you think I should add somethign like that?

Reply:
Ah, OK! I thought I was missing something. :-)

I don't think we need to block this PR -- we can definitely iterate on it later.

My preference, since you confirmed that callers would need to use a pinned box anyway, would be to just have have `new_pinned()` and remove `new()` until we have a way to actually use it (same for miscdev). Meanwhile, we can keep the internal logic even if it could be simplified, since we are still expecting to iterate on the API to support using the "module type" data (please create an issue so that we don't forget about this).

Reply:
> My preference, since you confirmed that callers would need to use a pinned box anyway, would be to just have `new_pinned()` and remove `new()` until we have a way to actually use it (same for miscdev).

I feel like we're having the same discussion we had with miscdev :)

KernelModule initialisation not playing well with pinning doesn't mean that `new` is useless. For example, KernelModule may have a heap-allocated object it points to (or a collection of such heap-allocated objects), and this object has a field that holds a registration object: in such a case, `new` is needed; if we were to only only have `new_pinned`, we would force clients to have an extra allocation.

In fact, this is how I use `miscdev` in binder, so please don't remove `new` from there.

Reply:
True, I was thinking about direct members! Note I only proposed to remove `new()` assuming it wasn't possible able to be used -- it is the first time I play with `Pin`, which is why I was asking whether it was true that there was no actual way to take advantage of `new()` :)